walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -49,93 +58,120 @@
-    }
-
-    int ret = binarySearchOverlappingTimeRange(mandatory, midIndex + 1, endIndex, startOfOptional);
-
-    return ret == -1 ? midIndex : ret;
+    return finalTimeRanges.size() != 0 ? finalTimeRanges : mandatory;
   }
 
-
   /*
-   * Method to get available times from sorted events
-   * Does this by removing events where there's no attendees, then merging events
-   * which share some time. Finally the gaps are the result.
+   * Removes events in which no attendees intersect between the event and requested attendees
    * 
-   * Runtime: O(n)
-   *
+   * Runtime O(n*q) where 'n' is the length of events and 'q' is the length of the attendees
+   * 
+   * @param attendes    A hashset of all applicable attendees
+   * @param events      A sorted list of the events to merge
+   * 
+   * return   A list of events where all events have at least one attendee from the set
+   *  
   */
-  public ArrayList<TimeRange> availableTimesFromSortedEvents(HashSet<String> attendees, ArrayList<Event> events, MeetingRequest request) {
-    ArrayList<TimeRange> availableTimes = new ArrayList<>();
-    ArrayList<Event> sequentialEvents = new ArrayList<>(events);
+  private LinkedList<Event> removeEventsWithNoAttendees(HashSet<String> attendees, LinkedList<Event> events) {
+    LinkedList<Event> sequentialEvents = new LinkedList<>(events);
 
-    // Removes events in which no attendees intersect between the event and requested attendees
-    for (int i = 0; i < sequentialEvents.size(); ++i) {
+    for (int i = 0; i < sequentialEvents.size(); ++i) { // n
       HashSet<String> attendeesForEvent = new HashSet<String>(sequentialEvents.get(i).getAttendees());
 
-      attendeesForEvent.retainAll(attendees);
+      boolean shareNoElement = Collections.disjoint(attendeesForEvent, attendees); // q
 
-      if (attendeesForEvent.size() <= 0) {
+      if (shareNoElement) {
         sequentialEvents.remove(i);
         --i;
       }
     }
-    
 
-    // Merges events which share some time
-    // Ex:       |---A---|
-    //                |--B--|
-    // Becomes:  |----------| <- MERGED_EVENT
-    for (int i = 0; i < sequentialEvents.size() - 1;) {
+    return sequentialEvents;
+  }
+
+  /*
+   * Method to merge overlapping events from a sorted array
+   * Ex:       |---A---|
+   *                |--B--|
+   * Becomes:  |----------| <- MERGED_EVENT 
+   * 
+   * Runtime: O(n) where 'n' is the length of the events
+   *
+   * @param attendes    A hashset of all applicable attendees
+   * @param events      A sorted list of the events to merge
+   * 
+   * return   A list of sorted and non overlapping sorted events
+   * 
+  */
+  private LinkedList<Event> sortedEventsIntoNonOverlappingSortedEvents(HashSet<String> attendees, LinkedList<Event> events) {
+    LinkedList<Event> sequentialEvents = new LinkedList<>(events); // n    
+
+    for (int i = 0; i < sequentialEvents.size() - 1; ++i) { // n
       Event eventToCheck = sequentialEvents.get(i);
```

Comment:
Ah so, a little tricky here, but .get(i) for a LinkedList is going to be `O(n)` because LinkedLists are not built for performant random indexing. You can use an iterator to avoid calling get() here: https://www.tutorialspoint.com/iterate-through-a-linkedlist-using-an-iterator-in-java

It's pretty similar to a C++ iterator, if you're familiar with those.